lib/Backends/CPU/libjit/libjit.cpp

```diff
@@ -1514,13 +1518,26 @@ int8_t libjit_element_cmp_lt_kernel_i8(dim_t idx, const int8_t *LHS,
   return libjit_scale_i32i8(lhs, pre, post, scale, 0) < rhs ? 1 : 0;
 }
 
-// tanh cannot be vectorized by LLVM yet. Therefore we use the following
+// Tanh cannot be vectorized by LLVM yet. Therefore we use the following
 // formula instead: 1 - 2 / (exp(x * 2) + 1), which is also used by Caffe2 and
 // provides a good accuracy.
 // Once LLVM supports the vectorization of tanh, we can replace this
 // approximation by a direct tanh call.
-DEFINE_DATA_PARALLEL_KERNEL(libjit_tanh_kernel_f, float,
-                            1 - 2 / (expf(LHS[idx] * 2) + 1))
+// When the LIBJIT compile option "-ffast-math" is enabled the intermediate
+// computation expf(x) for Tanh operator is not handled properly for very
+// large positive values which results in NaN values for the Tanh output.
+// Therefore when the "-ffast-math" is enabled we compute the Tanh such that
+// we avoid computing large values for the "expf" function.
+#ifdef FFAST_MATH
+DEFINE_DATA_PARALLEL_KERNEL_FUNC(libjit_tanh_kernel_f) {
+  float x = LHS[idx];
+  return x < 0 ? (+1 - 2 / (expf(+2 * x) + 1)) : (-1 + 2 / (expf(-2 * x) + 1));
```

Comment:
Did you try any other implementations here to see if it makes any perf difference? Not sure if it'll make a difference, but since tanh is an odd function you could try keeping the tanh calc the same and just always make x negative/invert it back. You could benchmark it to see. Also could look at `--dump-llvm-ir` / `--dump-llvm-asm` to see how well it's vectorized.

Reply:
@jfix71 You were right. The branching logic has some performance degradation.
I optimized them to not use branching and I obtained the following results.
- I tested the two kernels with a vector with 10^8 float values
- For **Tanh** the performance goes like this:
  - Before this PR: **1247 ms**
  - Branching: **2300 ms**
  - Optimized: **1260 ms**
- For **Sigmoid** the performance goes like this:
  - Before this PR: **1230 ms**
  - Branching: **2300 ms**
  - Optimized: **1256 ms**

So as you can see the latest optimized versions are very close to the previous.
The CPU tests pass (some CI Interpreter tests are broken by some latest commits).
Can we have this landed then?
Thanks!